### PR TITLE
PS4 support

### PR DESCRIPTION
**Pull request by compnerd**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by jrose-apple** [comment]
cc @gribozavr, who's shepherded most new platforms.

How do PS4 version numbers work? It does seem to make sense that PS3 and PS4 are completely separate OSs, but…

**Comment by gribozavr** [comment]
Please add a test like `./test/Parse/ConditionalCompilation/armAndroidTarget.swift`.

I think every PS* platform should be treated as a separate target; for the existing platforms we know that every PS* target runs a unique OS, and it is ABI-incompatible and probably ISA-incompatible to every other PS* target.  Even if the new one can run the programs for the previous one, it is typically an emulation mode that for all compilation concerns looks like the previous environment.

File: test/Parse/ConditionalCompilation/x86_64PS4Target.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// RUN: %swift -parse %s -verify -target x86_64-scei-ps4 -disable-objc-interop -parse-stdlib
+// RUN: %swift-ide-test -test-input-complete -source-filename %s -target x86_64-scei-ps4
+
+#if os(FreeBSD)
+// This block should not parse.
+// os(FreeBSD) does not imply os(PS4)
+let i : Int = "Hello"
```
Comment (gribozavr) [inline test/Parse/ConditionalCompilation/x86_64PS4Target.swift]: There should be no space before ":".

**Comment by compnerd** [comment]
Actually, lets wait for https://github.com/apple/swift/pull/3218 so that I can rebase on that and fix the linker directive handling.  But, Ive added the test, any other review comments would be appreciated so that I can fix it up in one go.

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
@compnerd Please take a look at the CI issues:

```
/home/buildnode/jenkins/workspace/swift-PR-Linux/swift/validation-test/StdlibUnittest/PS4.swift:7:24: error: type 'OSVersion' has no member 'PS4'
```

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
@compnerd

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb:1616:35: error: use of unresolved identifier 'reasaon'
      return "ps4Any(*, reason: \(reasaon))"
                                  ^~~~~~~
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb:1615:30: note: did you mean 'reason'?
    case .ps4Any(reason: let reason):
```

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb [hunk 2]
```diff
@@ -1605,6 +1612,9 @@ public enum TestRunPredicate : CustomStringConvertible {
     case .freeBSDAny(reason: let reason):
       return "freeBSDAny(*, reason: \(reason))"
 
+    case .ps4Any(reason: let reason):
+      return "ps4Any(*, reason: \(reasaon))"
```
Comment (pzmudzinski) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: typo in `reasaon`

**Comment by compnerd** [comment]
The test needs to be fixed, and I believe I need to hunt down one more issue in the linker option emission for PS4 work to be ready to merge.

**Comment by gribozavr** [comment]
@compnerd Sounds good.  Just @-mention me when you are ready.

**Comment by compnerd** [comment]
@swift-ci Please test

**Comment by compnerd** [comment]
@gribozavr I think that all the swift side of things are complete.

**Comment by compnerd** [comment]
@gribozavr if nothing else remains to be addressed, I think this can be merged.

File: lib/IRGen/IRGenModule.cpp [hunk 3]
```diff
@@ -796,6 +796,15 @@ llvm::SmallString<32> getTargetDependentLibraryOption(const llvm::Triple &T,
...
       buffer += '"';
+  } else if (T.isPS4()) {
+    bool quote = library.find(' ') != StringRef::npos;
+
+    buffer += "\01";
+    if (quote)
+      buffer += '"';
+    buffer += library;
+    if (quote)
+      buffer += '"';
```
Comment (gribozavr) [inline lib/IRGen/IRGenModule.cpp]: This matches what Clang does in its `PS4TargetCodeGenInfo`, so LGTM.

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by CodaFi** [comment]
@compnerd Do you know how the PS4 represents process arguments?

**Pull request merged**